src/engine/thread.js

```diff
@@ -27,92 +33,70 @@ class _StackFrame {
         this.warpMode = warpMode;
 
         /**
-         * Reported value from just executed block.
-         * @type {Any}
-         */
-        this.justReported = null;
-
-        /**
-         * The active block that is waiting on a promise.
-         * @type {string}
-         */
-        this.reporting = '';
-
-        /**
-         * Persists reported inputs during async block.
+         * Procedure parameters.
          * @type {Object}
          */
-        this.reported = null;
+        this.params = params;
 
         /**
-         * Name of waiting reporter.
-         * @type {string}
-         */
-        this.waitingReporter = null;
-
-        /**
-         * Procedure parameters.
+         * A context passed to block implementations.
          * @type {Object}
          */
-        this.params = null;
+        this.executionContext = {};
 
         /**
-         * A context passed to block implementations.
-         * @type {Object}
+         * Has this frame changed and need a reset?
+         * @type {boolean}
          */
-        this.executionContext = null;
+        this.needReset = false;
     }
 
     /**
-     * Reset all properties of the frame to pristine null and false states.
-     * Used to recycle.
+     * Reset some properties of the frame to default values. Used to recycle.
      * @return {_StackFrame} this
      */
     reset () {
-
         this.isLoop = false;
-        this.warpMode = false;
-        this.justReported = null;
-        this.reported = null;
-        this.waitingReporter = null;
-        this.params = null;
-        this.executionContext = null;
+        this.executionContext = {};
+        this.needReset = false;
 
         return this;
     }
 
     /**
      * Reuse an active stack frame in the stack.
-     * @param {?boolean} warpMode defaults to current warpMode
-     * @returns {_StackFrame} this
      */
-    reuse (warpMode = this.warpMode) {
-        this.reset();
-        this.warpMode = Boolean(warpMode);
-        return this;
+    reuse () {
+        this.isLoop = false;
+        this.executionContext = {};
+        this.needReset = false;
     }
 
     /**
      * Create or recycle a stack frame object.
      * @param {boolean} warpMode Enable warpMode on this frame.
+     * @param {object}
      * @returns {_StackFrame} The clean stack frame with correct warpMode setting.
      */
-    static create (warpMode) {
-        const stackFrame = _stackFrameFreeList.pop();
-        if (typeof stackFrame !== 'undefined') {
-            stackFrame.warpMode = Boolean(warpMode);
+    static create (warpMode, params) {
+        const stackFrame = _stackFrameFreeList.pop() || null;
+        if (stackFrame !== null) {
+            stackFrame.warpMode = warpMode;
```

Comment:
should `warpMode` still be cast to Boolean in the RHS here? I think it can sometimes be a string...

Reply:
I was wondering about this too. Also because I wanted to know what ways warpMode may be set.

https://github.com/LLK/scratch-vm/blob/develop/src/engine/thread.js#L247

If this warpMode was a string that means its parent warpMode would be a string. That is true in this PR and in develop. The thread creating a stack frame sets the warpMode to its parent warpMode. If the parent stack frame's warpMode is a string ... then we are already running with a warpMode that is a string.

Outside of Thread I found one line that changes warpMode.

https://github.com/LLK/scratch-vm/blob/develop/src/engine/sequencer.js#L339

Sequencer is handling the warp value in case its a string. And pretty explicitly setting warpMode to true if the warp in mutation is true.

I don't think I was previously aware, but this means we only need to check and set a warpTimer in the beginning of Sequnecer.stepThread and in Sequencer.stepToProcedure after setting warpMode true. Once warpMode is true, its true until we pop to a stack frame where its false.